oximeter/impl/src/schema/codegen.rs

```diff
@@ -0,0 +1,392 @@
+/// target.
+pub fn use_timeseries(contents: &str) -> Result<TokenStream, MetricsError> {
+    let schema = load_schema(contents)?;
+    let latest = find_schema_version(schema.iter().cloned(), None);
+    let mod_name = quote::format_ident!("{}", latest[0].target_name());
+    let types = emit_schema_types(latest);
+    let func = emit_schema_function(schema.into_iter());
+    Ok(quote! {
+        pub mod #mod_name {
+            #types
+            #func
+        }
+    })
+}
+
+fn emit_schema_function(
+    list: impl Iterator<Item = TimeseriesSchema>,
+) -> TokenStream {
+    quote! {
+        pub fn timeseries_schema() -> Vec<::oximeter::schema::TimeseriesSchema> {
+            vec![
+                #(#list),*
+            ]
+        }
+    }
+}
+
+fn emit_schema_types(list: Vec<TimeseriesSchema>) -> TokenStream {
+    let first_schema = list.first().expect("load_schema ensures non-empty");
+    let target_def = emit_target(first_schema);
+    let metric_defs = emit_metrics(&list);
+    quote! {
+        #target_def
+        #metric_defs
+    }
+}
+
+fn emit_metrics(schema: &[TimeseriesSchema]) -> TokenStream {
+    let items = schema.iter().map(|s| emit_one(FieldSource::Metric, s));
+    quote! { #(#items)* }
+}
+
+fn emit_target(schema: &TimeseriesSchema) -> TokenStream {
+    emit_one(FieldSource::Target, schema)
+}
+
+fn emit_one(source: FieldSource, schema: &TimeseriesSchema) -> TokenStream {
+    let name = match source {
+        FieldSource::Target => schema.target_name(),
+        FieldSource::Metric => schema.metric_name(),
+    };
+    let struct_name =
+        quote::format_ident!("{}", format!("{}", heck::AsPascalCase(name)));
+    let field_defs: Vec<_> = schema
+        .field_schema
+        .iter()
+        .filter_map(|s| {
+            if s.source == source {
+                let name = quote::format_ident!("{}", s.name);
+                let type_ = emit_rust_type_for_field(s.field_type);
+                let docstring = s.description.as_str();
+                Some(quote! {
+                    #[doc = #docstring]
+                    pub #name: #type_
+                })
+            } else {
+                None
+            }
+        })
+        .collect();
+    let (oximeter_trait, maybe_datum, type_docstring) = match source {
+        FieldSource::Target => (
+            quote! {::oximeter::Target },
+            quote! {},
+            schema.description.target.as_str(),
+        ),
+        FieldSource::Metric => {
+            let datum_type = emit_rust_type_for_datum_type(schema.datum_type);
+            (
+                quote! { ::oximeter::Metric },
+                quote! { pub datum: #datum_type },
+                schema.description.metric.as_str(),
+            )
+        }
+    };
+    quote! {
+        #[doc = #type_docstring]
+        #[derive(Debug, Clone, #oximeter_trait, PartialEq)]
+        pub struct #struct_name {
+            #(#field_defs),*
+            #maybe_datum
```

Comment:
Is this right? I *think* that `#( #field_defs ),*` will omit the trailing comma. So if we have one or more items in `field_defs` and then `maybe_datum` is `Some(_)` we'd have:

```rust
pub struct Foo {
    pub x: Y
    pub datum: D
}
```

I'd suggest:


```suggestion
            #( #field_defs, )*
            #maybe_datum ,
```

Reply:
Nice catch, thanks. This does look like a bug that works by accident. The one set of metrics I'm moving now all have no fields themselves, only those from the target. So we never hit the case you describe. I'll add tests for that stuff now.

Reply:
I added tests that catch this and fixed it up in 7cefed7a7